### PR TITLE
791-Define function for mode-based imputation

### PR DESCRIPTION
**Pull request by Ohi89**
## Pre-review checklist

I have ensured that:

- [x] Code runs successfully in the development environment.
- [x] New code is tested to prove that it works. If not, justify reasoning.
- [x] New code is documented using [type hinting](https://realpython.com/lessons/type-hinting/) and following the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring format.
- [x] New code follows the project naming standards.

## Post-review checklist

As I merge these changes, I have:

- [ ] Updated the related Jira ticket, including approximate time spent (including review).
- [ ] Moved the related Jira ticket to Done.

### CODE REVIEW

File: tests/impute/test_impute_mode.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+def test_impute_distribution(spark_session):
+
+    expected_df = spark_session.createDataFrame(
+        data=[
+            # current stata uses 1, 2 for "male", "female"
+            ("E12000007", "white", "male", None),  # proportion white-male = 1.0 therefore all rnd < 1.0
+            ("E12000007", "white", "male", None),
+            ("E12000007", None, None, "white"),
+            ("E12000007", "white", None, None),
```
Comment (jonathonmellor) [inline tests/impute/test_impute_mode.py:14]: The grouping needs to be done by household, so the test data needs a household indicator, the GOR9D is not needed.

**Review by jonathonmellor — CHANGES_REQUESTED** [review]
Code does not meet the imputation spec (incorrect columns grouped by, imputation done on population as a whole, not via the households, hard coded columns)

Code does not follow the convention of other functions for grouping.

The test used is inappropriate for this imputation as the function does not meet the spec. The test data needs to be related to the task/specific behavior needed, not used directly from the 792 imputation approach.

File: cishouseholds/impute.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    reference_column
+        The column for which imputation values will be calculated
+
+    Notes
+    -----
+    Function provides a column value for each record that needs to be imputed.
+    Where the value does not need to be imputed the column value created will be null.
+    """
+    # groupby count() after get all not nulls
+    df_notnulls = df.filter(F.col(reference_column).isNotNull()).groupby(df.white_group).count()
```
[the remark this answers is not in the document]
Comment (Ohi89) [reply cishouseholds/impute.py:23]: i dont know how i didn't see it was hardcoded sorry

[the remark this answers is not in the document]
Comment (jonathonmellor) [reply cishouseholds/impute.py:23]: Easily done!! No problem, this is what review's for 😄

File: cishouseholds/impute.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+    # dump maximum value
+    mode_imput = df_notnulls.agg({reference_column: "max"}).collect()[0][0]
```
[the remark this answers is not in the document]
Comment (Ohi89) [reply cishouseholds/impute.py:26]: This is sorted now

File: cishouseholds/impute.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ohi89) [reply cishouseholds/impute.py:23]: I used window functions but for some of the steps I have to use group by and usage of joins which might not be consistent but that's due to the specific logic with a few exceptions.

File: tests/impute/test_impute_mode.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+from cishouseholds.impute import calculate_imputation_from_mode
+
+
+def test_impute_distribution(spark_session):
+
+    expected_df = spark_session.createDataFrame(
+        data=[
+            # current stata uses 1, 2 for "male", "female"
+            ("E12000007", "white", "male", None),  # proportion white-male = 1.0 therefore all rnd < 1.0
```
[the remark this answers is not in the document]
Comment (Ohi89) [reply tests/impute/test_impute_mode.py:11]: I created dummy data specific to the mode imputation now.

File: tests/impute/test_impute_mode.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+            ("E12000007", "white", "male", None),
```
[the remark this answers is not in the document]
Comment (Ohi89) [reply tests/impute/test_impute_mode.py:12]: Done

**Comment by Ohi89** [comment]
Hi Jonathon, 

I have made a few changes to the code, sorted the hard-coded bits. Implemented a UDF (user defined function) for the specific logic that by using the built-in functions is not possible to do. I used window functions but at some stages I had to use groupby() and use several inner and outer joins.

We can have a look and i can explain what i have done whenever you want. There's quite a bit of code but this is due to my realization that the mode imputation task is more complicated than I thought.

**Comment by jonathonmellor** [comment]
Hi @Ohi89,

Sounds good! Happy to take a look if you could push it up.

Was something like https://stackoverflow.com/questions/65928710/pyspark-fill-nas-with-mode-of-column-based-on-aggregation-of-other-columns

your approach?

File: cishouseholds/impute.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ohi89) [reply cishouseholds/impute.py]: Sorted

File: tests/impute/test_impute_mode.py (hunk 1 above)
Comment (Ohi89) [reply tests/impute/test_impute_mode.py]: Grouping are possible to be selected. I have a purpose oriented data to test all the possible exceptions to make sure it passes correctly.

**Review by jonathonmellor — APPROVED** [review]

**Pull request merged**